prov/util/src/util_av.c

```diff
@@ -354,15 +355,32 @@ static void util_av_hash_remove(struct util_av_hash *hash, int slot, int index)
 int ofi_av_remove_addr(struct util_av *av, int slot, int index)
 {
 	struct util_ep *ep;
-	struct dlist_entry *av_entry;
 	int *entry, *next, i;
+	int ret;
 
 	if (index < 0 || (size_t)index > av->count) {
 		FI_WARN(av->prov, FI_LOG_AV, "index out of range\n");
 		return -FI_EINVAL;
 	}
 
 	fastlock_acquire(&av->lock);
+
+	/* This should stay at top */
+	dlist_foreach_container(&av->ep_list, struct util_ep, ep, av_entry) {
+		if (ep->cmap && ep->cmap->handles_av[index]) {
+			/* TODO this is not optimal. Replace this with something
+			 * more deterministic: delete handle if we know that peer
+			 * isn't actively communicating with us
+			 */
+			ret = ofi_cmap_move_handle_to_peer_list(ep->cmap, index);
+			if (ret) {
+				FI_WARN(av->prov, FI_LOG_DOMAIN,
+					"Unable to move handle to peer list\n");
+				return ret;
```

Comment:
A failure isn't likely, but maybe just delete the handle if we can't move it?

Reply:
But deleting would cause the issue we're trying to fix. Returning any error back to the user while trying to re-use the connection would be better, right? What do you think?

Reply:
I'm just not sure that the app can do anything with the failure.  Getting back a memory allocation failure as part of removing an address is odd.  If the alloc fails and we ignore it, what's the worst that happens?  Would this just keep the connection around floating in space until the app closes?

Reply:
The idea was to stop the app from doing anything more when the removal fails. The original issue occurs when the app removes the address and adds it again. Perhaps we could delete the handle but still return an error? I'm just trying to avoid the app hit the original issue. But this is a very corner case that's unlikely to happen anyway.

Reply:
I'm okay here, but here and below we return in case of failures holding a lock.